### PR TITLE
ready solution hw-17

### PR DESCRIPTION
**Pull request by RomanHreits**

### CODE REVIEW

File: src/main/java/core/basesyntax/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        if (str == null || str.isEmpty()) {
+            throw new IllegalArgumentException("String is empty!");
+        }
+        arithmeticOperation = operationsFactory(str);
+        double[] numbers = parserStringArrayToDoubleArray(str);
+        return arithmeticOperation.action(numbers[0], numbers[1]);
+    }
+
+    private double[] parserStringArrayToDoubleArray(String s) {
+        String[] split = s.split("[\\s\\W]+");
```
Comment (aleksandr-hrankin) [inline src/main/java/core/basesyntax/Calculator.java:37]: Я б здесь назвал не split а operands

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private static final double MULT_RESULT = 750.0;
+    private static final String POW = "2 ^ 7";
+    private static final double POW_RESULT = 128.0;
+    private static final String SUBTR = "33 - 8";
+    private static final double SUBTR_RESULT = 25.0;
+    private static final String DIVISION_BY_ZERO = "25 / 0";
+    private static final String STRING_WITHOUT_ARITHMETIC_CHARS = "25_.0";
+    private static final String STRING_WITH_MORE_THAN_TWO_NUMBERS = "25 + 30 + 45";
+    private static final double DELTA = 0.000_01;
+    private Calculator calculator = new Calculator();
```
Comment (aleksandr-hrankin) [inline src/test/java/core/basesyntax/CalculatorTest.java:22]: В чек листе говорят инициализировать класс через метод с аннотацией @Before или @BeforeClass

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    public void checkPassedArgumentWithWrongParameters() {
+        try {
+            calculator.calculate(STRING_WITHOUT_ARITHMETIC_CHARS);
+        } catch (IllegalArgumentException e) {
+            return;
+        }
+        fail("IllegalArgumentException was expected!!!");
```
Comment (aleksandr-hrankin) [inline src/test/java/core/basesyntax/CalculatorTest.java:46]: Как написали мне, напишу тебе)
We can use @test(expected = YourException.class) here

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    public void passEmptyStringToCalculateMethod() {
+        try {
+            calculator.calculate("");
+        } catch (IllegalArgumentException e) {
+            return;
+        }
+        fail("IllegalArgumentException was expected!!!");
+    }
```
Comment (aleksandr-hrankin) [inline src/test/java/core/basesyntax/CalculatorTest.java:57]: the same

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    public void passNullValueToCalculateMethod() {
+        try {
+            calculator.calculate(null);
+        } catch (IllegalArgumentException e) {
+            return;
+        }
+        fail("IllegalArgumentException was expected!!!");
+    }
```
Comment (aleksandr-hrankin) [inline src/test/java/core/basesyntax/CalculatorTest.java:67]: the same

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    public void checkDivisionByZero() {
+        try {
+            calculator.calculate(DIVISION_BY_ZERO);
+        } catch (ArithmeticException e) {
+            return;
+        }
+        fail("ArithmeticException was expected!!!");
+    }
```
Comment (aleksandr-hrankin) [inline src/test/java/core/basesyntax/CalculatorTest.java:77]: the same

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    public void checkPassedArgumentWithMoreThanTwoNumbers() {
+        try {
+            calculator.calculate(STRING_WITH_MORE_THAN_TWO_NUMBERS);
+        } catch (IllegalArgumentException e) {
+            return;
+        }
+        fail("IllegalArgumentException was expected!!!");
+    }
```
Comment (aleksandr-hrankin) [inline src/test/java/core/basesyntax/CalculatorTest.java:87]: the same

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 3 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:46]: Думав так зробити, але повторив за Богданом з лекції))

File: src/main/java/core/basesyntax/Calculator.java (hunk 1 above)
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java:37]: згоден

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 2 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:22]: Я думаю, що в даному прикладі, це вроді і не важливо...але умова є умова)

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 4 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:52]: done

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 5 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:57]: done

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 6 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:62]: done

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 7 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:67]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            calculator.calculate(DIVISION_BY_ZERO);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void checkPassedArgumentWithMoreThanTwoNumbers() {
+            calculator.calculate(STRING_WITH_MORE_THAN_TWO_NUMBERS);
+    }
+
+    @Test
+    public void isCorrectArithmeticOperations() {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:70]: let's add more test cases there: test with positives, negatives and zero
you may store input strings as local variables

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        assertEquals(ADD_RESULT, calculator.calculate(ADD), DELTA);
+        assertEquals(DIV_RESULT, calculator.calculate(DIV), DELTA);
+        assertEquals(MULT_RESULT, calculator.calculate(MULT), DELTA);
+        assertEquals(SUBTR_RESULT, calculator.calculate(SUBTR), DELTA);
+        assertEquals(POW_RESULT, calculator.calculate(POW), DELTA);
+    }
+}
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:77]: add a blank line

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Calculator.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package core.basesyntax;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Feel free to remove this class and create your own.
+ */
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java:8]: done

File: src/main/java/core/basesyntax/PowOperation.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax;
+
+public class PowOperation implements ArithmeticOperation {
+    @Override
+    public double action(double a, double b) {
+
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/PowOperation.java:6]: done

File: src/main/java/core/basesyntax/Calculator.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        double[] numbers = parserStringArrayToDoubleArray(str);
+        return arithmeticOperation.action(numbers[0], numbers[1]);
+    }
+
+    private double[] parserStringArrayToDoubleArray(String str) {
+        String[] numbers = str.split("[\\s\\W]+");
+        return new double[]{Double.valueOf(numbers[0]), Double.valueOf(numbers[1])};
+    }
+
+    private void operationsFactory(String str) {
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java:41]: done

File: src/main/java/core/basesyntax/Calculator.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    public double calculate(String str) {
+        if (str == null || str.isEmpty()) {
+            throw new IllegalArgumentException("String is empty!");
+        }
+        operationsFactory(str);
+        double[] numbers = parserStringArrayToDoubleArray(str);
+        return arithmeticOperation.action(numbers[0], numbers[1]);
+    }
+
+    private double[] parserStringArrayToDoubleArray(String str) {
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java:36]: done

File: src/main/java/core/basesyntax/Calculator.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        String[] numbers = str.split("[\\s\\W]+");
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java:37]: done

File: src/main/java/core/basesyntax/Calculator.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Feel free to remove this class and create your own.
+ */
+public class Calculator {
+
+    private ArithmeticOperation arithmeticOperation;
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java:11]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
+package core.basesyntax;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:6]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    private static final String MULT = "25 * 30";
+    private static final double MULT_RESULT = 750.0;
+    private static final String POW = "2 ^ 7";
+    private static final double POW_RESULT = 128.0;
+    private static final String SUBTR = "33 - 8";
+    private static final double SUBTR_RESULT = 25.0;
+    private static final String DIVISION_BY_ZERO = "25 / 0";
+    private static final String STRING_WITHOUT_ARITHMETIC_CHARS = "25_.0";
+    private static final String STRING_WITH_MORE_THAN_TWO_NUMBERS = "25 + 30 + 45";
+    private static final double DELTA = 0.000_001d;
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:22]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    private Calculator calculator;
+
+    @Before
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:25]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    public void init(){
+        calculator = new Calculator();
+    }
+    @Test
+    public void isWeCallTheRightClass() {
+        assertNull(calculator.getArithmeticOperation());
+        calculator.calculate(ADD);
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:32]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void checkPassedArgumentWithWrongParameters() {
+            calculator.calculate(STRING_WITHOUT_ARITHMETIC_CHARS);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void passEmptyStringToCalculateMethod() {
+            calculator.calculate("");
```
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:51]: DONE

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 9 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:77]: DONE

File: src/main/java/core/basesyntax/Calculator.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (RomanHreits) [reply src/main/java/core/basesyntax/Calculator.java]: done

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 8 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:72]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void checkPassedArgumentWithMoreThanTwoNumbers() {
+        String stringWithMoreThanTwoNumbers = "25 + 30 + 45";
+        calculator.calculate(stringWithMoreThanTwoNumbers);
+    }
+
+    @Test
+    public void isCorrectArithmeticOperations() {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:72]: nice work on the comments!
please create at least 5 different tests out of this one
to check 5 different operations (addition, subtraction and so on)

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 21 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java]: done

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void checkPassedArgumentWithMoreThanTwoNumbers() {
+        String stringWithMoreThanTwoNumbers = "25 + 30 + 45";
+        calculator.calculate(stringWithMoreThanTwoNumbers);
+    }
+
+    @Test
+    public void okSubtractOperations() {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:72]:
```suggestion
    public void checkSubtractOperations() {
```

**Review by KseniiaMakarova — APPROVED** [review]

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 22 above)
Comment (RomanHreits) [reply src/test/java/core/basesyntax/CalculatorTest.java:72]: прикол в тому що я так відразу і написав, але потім подумав що okSubtractOperations() буде краще))))
